### PR TITLE
manual typings for bindCallback.create

### PR DESCRIPTION
**Pull request by david-driscoll**
Comparing against generation of types to spur on discussion of the best approach.  I picked a single file here just as an example case.  See #1087

As a reference for TypeScript itself, these typings can't be simplified much more without losing the enhanced experience of operating against them.

https://github.com/Microsoft/TypeScript/issues/5453 may be the answer, not sure if it answers all the possible use cases.  It is currently on the road map for 2.0.

I realize there will be tons of opinions my _only_ goal is to make the TypeScript experience with RxJS5 as good as we have gotten the experience in RxJS4.

### CODE REVIEW

**Comment by blesh** [comment]
@david-driscoll I think you proposed this at an inopportune time (before the holidays).

Do my heavy TypeScript users have any opinions? @robwormald @jeffbcross @kwonoj

**Comment by kwonoj** [comment]
Discussion at the moment makes this discussion pause until https://github.com/ReactiveX/RxJS/pull/1077 is landed in codebase, to separate scope of changes. 

Shortly, I still need spend some time to review these changes.

**Comment by kwonoj** [comment]
Just to ensure, PR intended to be checked in is https://github.com/ReactiveX/RxJS/pull/1087, this PR shows handcrafted changes to compare with original PR.

**Comment by robwormald** [comment]
the better the types in the distribution, the better, as far as i'm concerned. i'll take an extra few kb's of npm download for better intellisense. Up to y'all in terms of maintainability.

**Comment by david-driscoll** [comment]
@kwonoj yes, I would prefer to check in is #1087, but I offered both to allow for comparing the difference between the two.

I have also rebased against master.

**Comment by david-driscoll** [comment]
@kwonoj / @luisgabriel 

After doing some work today, I'm thinking this branch might be the preferred method, I added some "unit tests" as well (basically to ensure that things compile as expected).

**Comment by kwonoj** [comment]
> this branch might be the preferred method, 

meaning between #1087 and this?

> I added some "unit tests" as well

Would you able to separate those? You may able to see there are pending PR in discussion https://github.com/ReactiveX/RxJS/pull/826 https://github.com/ReactiveX/RxJS/pull/667 for enabling test against type definitions, haven't concluded which way'll go. If your changes separated, that can be considerable option too.

**Comment by david-driscoll** [comment]
Correct, this over #1087, it's a little more manual work, but the quality will probably be better.

If you look at `bindCallback-tpec.ts` for example, it doesn't actually do anything meaningful, in terms of testing functionality, it just validates that interface defined for `bindCallback` will compile as expected with the correct type inferences, etc.

**Comment by kwonoj** [comment]
> If you look at bindCallback-tpec.ts for example,

yes, aware of those. Current discussion regarding test is about organization of test since it's testing definition only - so it'll be matter of trying to use existing PR's mechanism, or following your PR's mechanism to invoke those. Once it's decided, test can be added fairly quickly.

**Comment by david-driscoll** [comment]
@kwonoj done!

**Comment by kwonoj** [comment]
Thanks! I still need to look changes to add some inputs.. (my work-for-pay-bill blocks me, sorry for late response)

File: src/add/operator/combineLatest.ts [hunk 1]
```diff
@@ -5,6 +5,6 @@
 import {Observable} from '../../Observable';
 import {combineLatest} from '../../operator/combineLatest';
 
-Observable.prototype.combineLatest = combineLatest;
+Observable.prototype.combineLatest = <any>combineLatest;
```
Comment (kwonoj) [inline src/add/operator/combineLatest.ts]: this patch operator's now auto generated, so instead of changing this script need to be updated - is that correct?

File: src/observable/bindCallback.ts [hunk 2]
```diff
@@ -9,6 +9,24 @@ import {AsyncSubject} from '../subject/AsyncSubject';
 export class BoundCallbackObservable<T> extends Observable<T> {
   subject: AsyncSubject<T>;
 
+  /* tslint:disable:max-line-length */
+  static create<TResult>(callbackFunc: (callback: (result: TResult) => any) => any, selector?: void, scheduler?: Scheduler): () => Observable<TResult>;
```
Comment (kwonoj) [inline src/observable/bindCallback.ts]: My assumption is any similar operator would need same amount of declarations (i.e, `BoundNodeCallbackObservable`) - is that correct?

File: src/operator/combineLatest.ts [hunk 3]
```diff
@@ -31,3 +31,28 @@ export function combineLatest<T, R>(...observables: Array<Observable<any> |
 
   return new ArrayObservable(observables).lift<T, R>(new CombineLatestOperator<T, R>(project));
 }
+
+/* These types would be come standard types used through-out the library */
```
Comment (kwonoj) [inline src/operator/combineLatest.ts]: Means this might need to be exported into different module later in common usage?

File: src/operator/combineLatest.ts [hunk 4]
```diff
@@ -31,3 +31,28 @@ export function combineLatest<T, R>(...observables: Array<Observable<any> |
...
+export type ObservableOrPromise<T> = Observable<T> | Promise<T>;
+export type ArrayOrIterator<T> = Iterator<T> | ArrayLike<T> | Array<T>;
+export type ObservableInput<T> = Observable<T> | Promise<T> | Iterator<T> | ArrayLike<T>;
+
+/* tslint:disable:max-line-length */
+export interface CombineLatestDeclaration<T> {
```
Comment (kwonoj) [inline src/operator/combineLatest.ts]: (It's just naming) I'm currently mixing usage of term between `Declaration` with `Definition`, which one looks more intuitive name? or maybe `CombineLatestSignature`, `CombineLatestType`, etcs?

**Comment by kwonoj** [comment]
Let me ask few questions as 'newcomer' point of view, let's say I'm completely new to codebase and compose new operator require similar function signatures

- How do I determine if operator requires this 1 to n type definition or not?
- Where can I refer to create those signatures? such as base templates, examples, etcs
- How do I correctly verify once it's composed to send out PR? (Yes we do have https://github.com/ReactiveX/RxJS/issues/661, questions about 'how to compose those test')

There might be some other questions I could possible missing.

**Comment by kwonoj** [comment]
+ would you agree to close #1087 in case if this PR's expected to be checked in? (confirming if I understood correctly)

File: src/add/operator/combineLatest.ts (hunk 1 above)
Comment (luisgabriel) [inline src/add/operator/combineLatest.ts]: Is this change needed?

Comment (david-driscoll) [inline src/add/operator/combineLatest.ts]: yes, because `combineLatest` (the method) no longer has the same definition as `combineLatest` the prototype method.

I plan to update the script to add this in, and that's when I found out that the generation script didn't work the other day.

File: src/observable/bindCallback.ts (hunk 2 above)
Comment (david-driscoll) [inline src/observable/bindCallback.ts]: I don't know how important that is to be honest.  My feeling is we don't need to, because generally no one is going to use the Observable class, they're almost always going to go through the method.

File: src/operator/combineLatest.ts (hunk 3 above)
Comment (david-driscoll) [inline src/operator/combineLatest.ts]: Yeah, and we may need discussion of them as well.

File: src/operator/combineLatest.ts (hunk 4 above)
Comment (david-driscoll) [inline src/operator/combineLatest.ts]: Naming is hard, I just picked something at the time that didn't clash.

I kind of like `CombineLatestSignature` it's descriptive as to what this interface really is about.

**Comment by david-driscoll** [comment]
It only apply to methods that have a method that has some sort of applied / combined output or input.  Which is actually only a few, but they are the important ones to capture, because it makes type inference that much better.  Exaples are `combineLatest`, `zip`, `withLatestFrom`, etc.

The only other most common case is methods with an optional selector.  You'd want one overload without the selector, because generally those return the incoming type `T` in some form, and the one with the selector would return `R`.

File: src/add/operator/combineLatest.ts (hunk 1 above)
Comment (luisgabriel) [inline src/add/operator/combineLatest.ts]: Hmn. Makes sense.

File: src/operator/combineLatest.ts (hunk 4 above)
Comment (luisgabriel) [inline src/operator/combineLatest.ts]: +1 for `CombineLatestSignature`.

**Comment by luisgabriel** [comment]
I don't now what do you guys think but I'd prefer to split this into two separate commits: one for `biundCallback` and another for `combineLatest`.

File: src/observable/bindCallback.ts [hunk 5]
```diff
@@ -9,6 +9,24 @@ import {AsyncSubject} from '../subject/AsyncSubject';
...
+  static create<T, TResult>(callbackFunc: (v1: T, callback: (result: TResult) => any) => any, selector?: void, scheduler?: Scheduler): (v1: T) => Observable<TResult>;
+  static create<T, T2, TResult>(callbackFunc: (v1: T, v2: T2, callback: (result: TResult) => any) => any, selector?: void, scheduler?: Scheduler): (v1: T, v2: T2) => Observable<TResult>;
```
Comment (luisgabriel) [inline src/observable/bindCallback.ts]: I'd prefer to rename `TResult` to `R`. It's less verbose and keeps the same pattern we're using for the other operators.

File: src/observable/bindCallback.ts (hunk 2 above)
Comment (kwonoj) [inline src/observable/bindCallback.ts]: Would you elaborate bit more? does that mean this signature update is not even required by `Observable` should have correct signature?

**Comment by kwonoj** [comment]
> I'd prefer to split this into two separate commits: 

: +1 for those. Thanks for catching it.

**Comment by kwonoj** [comment]
@david-driscoll Reason I asked above questions are feeling those will be raised at some point when anyone try to make changes to contribute, tried to play advocate for those perspective. As same as some guide we do have for contribution or migration, I think these also need to be noted for further reference / maintenance purpose. (including above questions + list of operators have those signatures, etcs)

File: src/observable/bindCallback.ts (hunk 5 above)
Comment (kwonoj) [inline src/observable/bindCallback.ts]: +1 for now. I'm also quite familiar of `TResult` by living in the C# world, but let's make consistency first. If we'd like to have rename to this, can be discussed in separate PR.

File: src/operator/combineLatest.ts (hunk 4 above)
Comment (kwonoj) [inline src/operator/combineLatest.ts]: `CombineLatestSignature` seems feasible, let's go with that.

cvrjd2oukaahvmy

File: src/add/operator/combineLatest.ts (hunk 1 above)
Comment (david-driscoll) [inline src/add/operator/combineLatest.ts]: I've updated this in the last third commit as something separate from the other work. (Hence the larger number of files changed)

**Comment by david-driscoll** [comment]
Updated.

File: src/Observable.ts [hunk 6]
```diff
@@ -187,9 +194,7 @@ export class Observable<T> implements CoreOperators<T>  {
   bufferWhen: (closingSelector: () => Observable<any>) => Observable<T[]>;
   catch: (selector: (err: any, source: Observable<T>, caught: Observable<any>) => Observable<any>) => Observable<T>;
   combineAll: <R>(project?: (...values: Array<any>) => R) => Observable<R>;
-  combineLatest: <R>(...observables: Array<Observable<any> |
-                                     Array<Observable<any>> |
-                                     ((...values: Array<any>) => R)>) => Observable<R>;
+  combineLatest: CombineLatestSignature<this>;
```
Comment (kwonoj) [inline src/Observable.ts]: How do we assign updated signature of `bindCallback` here for static creation method? seems this PR doesn't changes it in `Observable`.

Comment (david-driscoll) [inline src/Observable.ts]: It's already bound, based on the static methods.

Comment (kwonoj) [inline src/Observable.ts]: Just checked again and you're correct, sorry for misunderstanding.

File: src/Observable.ts [hunk 7]
```diff
@@ -37,6 +37,13 @@ import {ErrorObservable} from './observable/throw';
 import {AjaxCreationMethod} from './observable/dom/ajax';
 import {WebSocketSubject} from './observable/dom/webSocket';
 
+import {CombineLatestSignature} from './operator/combineLatest';
+
+/* These types would be come standard types used through-out the library */
+export type ObservableOrPromise<T> = Observable<T> | Promise<T>;
```
Comment (kwonoj) [inline src/Observable.ts]: Let's separate this. Thinking about PR like https://github.com/ReactiveX/RxJS/pull/1175, I think there might be enough common usages for these types. Those implementation only might need import this definition, not whole `Observable`.

Comment (david-driscoll) [inline src/Observable.ts]: Separate commit or separate PR (to go first?)

**Comment by kwonoj** [comment]
Change looks good to me in general, (might) need to separate common type definitions.

Still, separate from this PR itself which won't block check in this, I'd like to discuss how to organize some kind of document (or tutorial) for this kind of operators anyone who want to contribute either creating new one or modifying existing one.

File: src/Observable.ts (hunk 7 above)
Comment (kwonoj) [inline src/Observable.ts]: As long as we keep in mind further refactoring to use this common type in existing codebase either way's fine. but separate commit would be easier for you I assume? :)

Comment (david-driscoll) [inline src/Observable.ts]: Not too worried about the work / effort, more just the ordering. :smile:

**Comment by kwonoj** [comment]
cc @blesh since this PR's nearly complete review, giving visibility before actual check in happens. In my perspective this change's are good and only waiting one small update but there could be other suggestions that I was missing about.

**Comment by david-driscoll** [comment]
@kwonoj added in the extra commit and have a PR for it, if we want a separate discussion.  See #1216

**Comment by masaeedu** [comment]
@david-driscoll Why the `<any>` casts in the patch files? Not having it serves as a nice check to ensure the implementation satisfies the interface.

File: src/add/operator/combineLatest.ts (hunk 1 above)
Comment (masaeedu) [inline src/add/operator/combineLatest.ts]: I'm not sure this should be needed. The implementation function should be assignable to the interface. If it doesn't do that, it means there's some overloads in the interface that the implementation isn't equipped to handle. 

In this case, the interface is advertising to users that they can pass in iterators, promises or arrays wherever observables are expected, however the function isn't written to expect these and will explode. This is causing the assignability error, which is a good thing.

Comment (masaeedu) [inline src/add/operator/combineLatest.ts]: I would recommend narrowing `ObservableInput<T>` to just `Observable<T>` until we can figure out a way to generically deal with iterators, arrays and promises in the actual implementations (which would be absolutely magical!)

**Comment by david-driscoll** [comment]
With the separated signature the implementation no longer satisfies the interface :cry:

**Comment by masaeedu** [comment]
Yeah, but if you change `ObservableInput<T>` from whatever it is now to just `Observable<T>`, it does. This is the reality of the implementations at the moment, but it would be amazing if we could update the implementations to deal with arraylikes, iterators and promises transparently. Possibly some kind of decorator for the arguments could be written.

**Comment by david-driscoll** [comment]
I hadn't noticed that point... I'll take a look tonight.

**Comment by kwonoj** [comment]
@david-driscoll , would you able to rebase this one? also do we have conclusions on `any` casting on patch operators?

**Comment by david-driscoll** [comment]
@kwonoj rebased!  I was able to eliminate the <any> calls, so that's already been removed from the changes.

**Comment by masaeedu** [comment]
@david-driscoll Right now the implementation for `combineLatest` accepts `Array<ObservableInput<any>>`. This won't work once the `lift` method is fixed, because you can't lift an `Observable<Observable<any> | Promise<any> | Iterator<any> | ArrayLike<any> ...>` through an `Operator<T, R>` and produce an `Observable<R>`. 

I'd suggest narrowing `ObservableInput` to `Observable` only for now, so that the API exposed to the user reflects what the operators can actually handle.

**Comment by david-driscoll** [comment]
@masaeedu I don't see any cosntraints in the code that would cause any compilation problems.  To validate I merged with your changes (which merge cleanly btw! :heart: git).

Do you have a specific example where you have seen issues pop in?

**Comment by masaeedu** [comment]
That's odd, it doesn't compile for me. The changes I'm making are to remove `T` from `Observable.lift` and removing `T` from `Operator.call`. Once you do that, you get an error on the last line of `combineLatest` explaining how `new ArrayObservable(observables).lift` expects an `Operator<Observable<any> | Promise<any> ...>`, which an `Operator<T, R>` does not satisfy. Could you try making those two changes and attempting a compile? I can provide a patch file if you want.

**Comment by david-driscoll** [comment]
Want to push your local branch and I'll check it out and see if I can see what you're seeing?

**Comment by masaeedu** [comment]
@david-driscoll Yup, here you go: https://github.com/masaeedu/RxJS/tree/typings-option-a. Look at the error for `combineLatest.ts`.

**Comment by david-driscoll** [comment]
Found the problem.

Your changes removed `<T, R>`

**Comment by david-driscoll** [comment]
Found the problem.

Your changes removed `<T, R>` from `new ArrayObservable(observables).lift(new CombineLatestOperator(project));` hence why I didn't the problem when merged. :smile: 

Now that said the real solution isn't actually anything that is wrong, the typings are just incorrect.

Currently ArrayObservable is typed like `ArrayObservable<Observable<any>>` or with my changes `ArrayObservable<ObservableInput<any>>`.  This isn't actually correct, instead ArrayObservable should container whatever type is contained by the inputs. (Observables, Promises, Iterables, Arrays, etc).

The best way I see to type around that is something like:

    new ArrayObservable<T>(<any[]>observables).lift<R>(new CombineLatestOperator<T, R>(project));

This produces the effect that we're looking for.   In call cases of `ArrayObservable`, should probably be typed similarly.

**Comment by masaeedu** [comment]
@david-driscoll You have to remove at least `T` from `.lift`, since there's only one free generic parameter in `lift` after the fix. Specifying `R` is unnecessary since it can be inferred from the type of the `Operator` being passed in, but even if you specify it this doesn't fix the problem (at least on my setup).

**Comment by masaeedu** [comment]
The reason I think the error is understandable and correct is that an `Operator<T, R>` can only turn an `Observable<T>` into an `Observable<R>`. The code does not exist to allow an operator to deal with `Promise`s or `Iterable`s or whatever else. With your typings, a user would assume they can pass promises or arrays into `combineLatest` where before `Observable`s were expected, but there is no additional code to deal with this and it will cause runtime errors.

**Comment by david-driscoll** [comment]
Two different problems.

1) The problem you noted was dealing with an `Observable` of `Observable<T>`, which should only be `Observable<T>`.  Also for the code itself `T` isn't really used, so just confuses matters to be honest.

2) Once you dig into `combineLatest-support.ts` you can see that it just shifts each value from the `ArrayObservable` into an array.  From there on complete it calls through to `subscribeToResult`, which handles among other things Observables, Promises, Arrays and Iterables.

**Comment by masaeedu** [comment]
Ah, I wasn't aware `subscribeToResult` transparently deals with other types of items. Sorry about that; I was assuming that because everything is typed to `Observable`, there was no code to deal with the other stuff. 

I actually have a version of `CombineLatestOperator` with updated typings [here](https://github.com/masaeedu/RxJS/blob/8a448ce26f8be52e9c09563162606b468cc1a80b/src/operator/combineLatest-support.ts), where the signature is:

```
class CombineLatestOperator<T extends Observable<U>, U, R> implements Operator<T, U[] | R>
```

I would just need to update this to:

```
CombineLatestOperator<T extends ObservableInput<U>, U, R> implements Operator<T, U[] | R>
```

I'll update stuff in my branch.

**Comment by kwonoj** [comment]
I've already marked this as lgtm & latest update removes `<any>` casting resolves some of questions we had. I'll check this in around this afternoon if there's any other suggestions around.

**Comment by blesh** [comment]
It looks like there are some merge conflicts... @david-driscoll can you rebase this?

@kwonoj, feel free to merge this whenever you think is appropriate if you feel this is an improvement to our typings.

**Comment by kwonoj** [comment]
Will do. I've already completed reviewing this and so far I believe good to check in. I'll do it myself once it's rebased.

**Comment by masaeedu** [comment]
Is this going to be a pattern that is followed uniformly for all the operators or just for the ones with complicated type signatures?

**Comment by kwonoj** [comment]
@masaeedu Not all of operators. There are some operators allows these kind of n-numbers of parameter s, those need to be updated like zip, withlatestfrom (https://github.com/ReactiveX/RxJS/issues/1208 , https://github.com/ReactiveX/RxJS/issues/1209)

**Comment by kwonoj** [comment]
@david-driscoll , I'll check this in once it's rebased. (Gentle reminder, not poking :wink: )

File: src/operator/combineLatest.ts [hunk 8]
```diff
@@ -48,8 +69,25 @@ export function combineLatest<T, R>(...observables: Array<Observable<any> |
  * @returns {Observable} an observable of other projected values from the most recent values from each observable, or an array of each of
  * the most recent values from each observable.
  */
-export function combineLatestStatic<T, R>(...observables: Array<any | Observable<any> |
-                                                    Array<Observable<any>> |
+/* tslint:disable:max-line-length */
+export function combineLatestStatic<T>(v1: ObservableInput<T>): Observable<[T]>;
```
Comment (david-driscoll) [inline src/operator/combineLatest.ts]: Since these have been merged, I added definitions for `combineLatestStatic` that mirror the signatures.

**Comment by david-driscoll** [comment]
@kwonoj rebased.

**Comment by kwonoj** [comment]
Thanks! I'll check this in.

**Comment by kwonoj** [comment]
Merged with 7c9547a, appreciate for huge effort @david-driscoll :smile:
